### PR TITLE
Stop policy on last PhaseCompleteStep instead of TerminalPolic…

### PR DESCRIPTION
**Pull request by dakrone**
Currently when an ILM policy finishes its execution, the index moves into the `TerminalPolicyStep`,
denoted by a completed/completed/completed phase/action/step lifecycle execution state.

This commit changes the behavior so that the index lifecycle execution state halts at the last
configured phase's `PhaseCompleteStep`, so for instance, if an index were configured with a policy
containing a `hot` and `cold` phase, the index would stop at the `cold/complete/complete`
`PhaseCompleteStep`. This allows an ILM user to update the policy to add any later phases and have
indices configured to use that policy pick up execution at the newly added "later" phase. For
example, if a `delete` phase were added to the policy specified about, the index would then move
from `cold/complete/complete` into the `delete` phase.

Relates to #48431

### CODE REVIEW

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java [hunk 1]
```diff
@@ -341,7 +342,7 @@ public void testAliasReplicatedOnShrink() throws Exception {
             assertBusy(() -> assertOK(client().performRequest(new Request("HEAD", "/" + shrunkenIndexName + "/_alias/" + indexName))));
 
             // Wait for the index to complete its policy
-            assertBusy(() -> assertILMPolicy(client(), shrunkenIndexName, policyName, "completed", "completed", "completed"));
+            assertBusy(() -> assertILMPolicy(client(), shrunkenIndexName, policyName, null, "complete", "complete"));
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java]: I'm probably missing something, but why is the phase `null` in these CCR cases? Should this be the name of the phase that was completed? (ie. warm)

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java [hunk 2]
```diff
@@ -78,7 +78,8 @@ public static void validateTransition(IndexMetaData idxMeta, Step.StepKey curren
                 "], currently: [" + realKey + "]");
         }
 
-        if (stepRegistry.stepExists(indexPolicySetting, newStepKey) == false) {
+        // Always allow moving to the terminal step, even if it doesn't exist in the policy
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java]: should we document this as it seems like a behavioural change for the `move-to-step` api?

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java [hunk 3]
```diff
@@ -374,80 +374,7 @@ public void testMasterDedicatedDataDedicated() throws Exception {
         assertBusy(() -> {
             LifecycleExecutionState lifecycleState = LifecycleExecutionState.fromIndexMetadata(client().admin().cluster()
                 .prepareState().execute().actionGet().getState().getMetaData().index("test"));
-            assertThat(lifecycleState.getStep(), equalTo(TerminalPolicyStep.KEY.getName()));
-        });
-    }
-
-    public void testMasterFailover() throws Exception {
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java]: was the deletion of this test intentional? if so shall we do it in a different commit so we have some semantics as to why it was removed?

**Review by andreidan — COMMENTED** [review]
Thanks for working on this @dakrone This looks great overall, I left a few questions but I think this is mostly ready

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java]: `null` just means that the matcher should not try to match the phase. The reason for that is that in these tests there is a `randomBoolean()` check that randomly uses either a lone `warm` phase or adds a `cold` phase, so we don't assert whether it's always `cold` or `warm` (it's only important that it reaches the `complete` step)

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java (hunk 2 above)
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java]: I looked through the documentation and didn't see any actual documentation about the terminal step already, so I'm not sure we want to get into that deep of technical behavior in our docs? What do you think?

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java (hunk 3 above)
Comment (dakrone) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java]: This was a test that didn't really do much outside of what we already have tested elsewhere, in addition, it was using a fake step that was near impossible to make work with the way that phases now operate, so I removed it.

I don't think a separate commit is that useful, since this ends up squashed into a single commit anyway?

**Comment by probakowski** [comment]
LGTM, thanks @dakrone!

**Review by probakowski — APPROVED** [review]

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java (hunk 1 above)
Comment (andreidan) [reply x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java]: Ah gotcha, the implementation skips that assertion if the phase is `null`. Thanks for the clarification, this makes sense

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java (hunk 2 above)
Comment (andreidan) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java]: Ah, sure, let's not mention it then

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java (hunk 3 above)
Comment (andreidan) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java]: Ah, yeah, you're right (not sure how should we signal the reasoning behind these changes for future reference - commit message?)

**Review by andreidan — APPROVED** [review]
LGTM thanks @dakrone
